# Importing Modules, defining functions

In [25]:
import tkinter as tk
from tkinter import filedialog
import numpy as np
import pandas as pd

def findlinevaldict(lines, valname):
    """
    Find all lines in lines (list) with valname (string) and
    return a dict of IDs and valuse coming after valname
    """
    alllineswithvalname = {}
    for line in lines:
        if valname in line:
            pos = line.find(valname) + len(valname)
            key, value = line.split()[0], [float(x) for x in line[pos:].split()]
            alllineswithvalname[key] = value
    return alllineswithvalname

def findlinevallist(lines, valname):
    """
    Find all lines in lines (list) with valname (string) and
    return a list of IDs and valuse coming after valname
    """
    alllineswithvalname = list()
    for line in lines:
        if valname in line:
            pos = line.find(valname) + len(valname)
            alllineswithvalname.append([line.split()[0], [float(x) for x in line[pos:].split()]])
    alllineswithvalname.sort()
    return alllineswithvalname

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

print('done')

done


# Getting file name

## Via file dialog

In [34]:
#if the file is not next to the notebook: getting file path 
root = tk.Tk()

fileName = filedialog.askopenfilename()

root.destroy()
del root

print(fileName)

2025-02-25 10:29:10.431 Python[69196:3011517] +[CATransaction synchronize] called within transaction
2025-02-25 10:29:10.482 Python[69196:3011517] +[CATransaction synchronize] called within transaction
2025-02-25 10:29:12.308 Python[69196:3011517] +[CATransaction synchronize] called within transaction


/Users/santapile/Santas/Projects/MaMMoS/dataExamples/UU/UU-sets/Mn2CrB4/GS/x/out_last


## Or if the file is next to the notebook

In [ ]:
#if the file is next to the notebook: 
# fileName = 'out_last'

# Reading file, saving parameters

In [35]:
# reading file into lines
with open(fileName, 'rt') as f:
        lines = f.read().splitlines()

# create dicts with orbital IDs and values of corresponding
# total moment and its direction (+/-1)
totmomentsD = findlinevaldict(lines, 'Total moment [J=L+S] (mu_B):')
dirofJD = findlinevaldict(lines, 'Direction of J (Cartesian):')

totmomentsL = findlinevallist(lines, 'Total moment [J=L+S] (mu_B):')
dirofJL = findlinevallist(lines, 'Direction of J (Cartesian):')


paramsmerged = {}
for key in totmomentsD.keys():
    paramsmerged[key] = totmomentsD[key] + dirofJD[key]

## Question #1: Why doubled?

In [36]:
totmomentsL

[['ID:0100010100-o', [0.002103611, 0.002103611]],
 ['ID:0100010100-o', [0.002103611, 0.002103611]],
 ['ID:0101010100-o', [0.001853875, 0.001853875]],
 ['ID:0101010100-o', [0.001853875, 0.001853875]],
 ['ID:0102010100-o', [0.166403, 0.166403]],
 ['ID:0102010100-o', [0.166403, 0.166403]],
 ['ID:0200010100-o', [0.01078338, 0.01078338]],
 ['ID:0200010100-o', [0.01078338, 0.01078338]],
 ['ID:0201010100-o', [0.03702574, 0.03702574]],
 ['ID:0201010100-o', [0.03702574, 0.03702574]],
 ['ID:0300010100-o', [0.01078338, 0.01078338]],
 ['ID:0300010100-o', [0.01078338, 0.01078338]],
 ['ID:0301010100-o', [0.03702562, 0.03702562]],
 ['ID:0301010100-o', [0.03702562, 0.03702562]],
 ['ID:0400010100-o', [0.004421223, 0.004421223]],
 ['ID:0400010100-o', [0.004421223, 0.004421223]],
 ['ID:0401010100-o', [0.01127024, 0.01127024]],
 ['ID:0401010100-o', [0.01127024, 0.01127024]],
 ['ID:0500010100-o', [0.004421223, 0.004421223]],
 ['ID:0500010100-o', [0.004421223, 0.004421223]],
 ['ID:0501010100-o', [0.01127029

In [37]:
dirofJL

[['ID:0100010100-o', [-1.0, 5.47e-06, -3.26e-06]],
 ['ID:0100010100-o', [-1.0, 5.47e-06, -3.26e-06]],
 ['ID:0101010100-o', [-1.0, -2.01e-06, 2.543e-05]],
 ['ID:0101010100-o', [-1.0, -2.01e-06, 2.543e-05]],
 ['ID:0102010100-o', [-1.0, 2.178e-05, 7.473e-05]],
 ['ID:0102010100-o', [-1.0, 2.178e-05, 7.473e-05]],
 ['ID:0200010100-o', [-1.0, -4.37e-06, -4.63e-06]],
 ['ID:0200010100-o', [-1.0, -4.37e-06, -4.63e-06]],
 ['ID:0201010100-o', [-1.0, -4.2e-07, 8.96e-06]],
 ['ID:0201010100-o', [-1.0, -4.2e-07, 8.96e-06]],
 ['ID:0300010100-o', [-1.0, -4.38e-06, -4.7e-06]],
 ['ID:0300010100-o', [-1.0, -4.38e-06, -4.7e-06]],
 ['ID:0301010100-o', [-1.0, -3e-07, 5.34e-06]],
 ['ID:0301010100-o', [-1.0, -3e-07, 5.34e-06]],
 ['ID:0400010100-o', [-1.0, -3.06e-06, -6.31e-06]],
 ['ID:0400010100-o', [-1.0, -3.06e-06, -6.31e-06]],
 ['ID:0401010100-o', [-1.0, -7e-06, 1.213e-05]],
 ['ID:0401010100-o', [-1.0, -7e-06, 1.213e-05]],
 ['ID:0500010100-o', [-1.0, -3.11e-06, -6.32e-06]],
 ['ID:0500010100-o', [-1.0, -3.11e

## Question #2: How do we calculate total magnetic moment? (Does direction matter?)

In [38]:
df = pd.DataFrame.from_dict(paramsmerged, orient='index',
                            columns = ['J=L+S (Cartesian)',
                                       'J=L+S (Spin axis)',
                                       'Direction of J (x?)',
                                       'Direction of J (y?)',
                                       'Direction of J (z?)'])
df

J=L+S (Cartesian)  J=L+S (Spin axis)  Direction of J (x?)  \
ID:0102010100-o           0.166403           0.166403                 -1.0   
ID:0101010100-o           0.001854           0.001854                 -1.0   
ID:0100010100-o           0.002104           0.002104                 -1.0   
ID:0201010100-o           0.037026           0.037026                 -1.0   
ID:0200010100-o           0.010783           0.010783                 -1.0   
ID:0301010100-o           0.037026           0.037026                 -1.0   
ID:0300010100-o           0.010783           0.010783                 -1.0   
ID:0401010100-o           0.011270           0.011270                 -1.0   
ID:0400010100-o           0.004421           0.004421                 -1.0   
ID:0501010100-o           0.011270           0.011270                 -1.0   
ID:0500010100-o           0.004421           0.004421                 -1.0   
ID:0602010100-o           1.775499           1.775499                  1.0   
ID:0601010100-o           0.005087           0.005087                  1.0   
ID:0600010100-o           0.005813           0.005813                  1.0   
ID:0702010100-o           1.775499           1.775499                  1.0   
ID:0701010100-o           0.005087           0.005087                  1.0   
ID:0700010100-o           0.005813           0.005813                  1.0   

                 Direction of J (y?)  Direction of J (z?)  
ID:0102010100-o         2.178000e-05         7.473000e-05  
ID:0101010100-o        -2.010000e-06         2.543000e-05  
ID:0100010100-o         5.470000e-06        -3.260000e-06  
ID:0201010100-o        -4.200000e-07         8.960000e-06  
ID:0200010100-o        -4.370000e-06        -4.630000e-06  
ID:0301010100-o        -3.000000e-07         5.340000e-06  
ID:0300010100-o        -4.380000e-06        -4.700000e-06  
ID:0401010100-o        -7.000000e-06         1.213000e-05  
ID:0400010100-o        -3.060000e-06        -6.310000e-06  
ID:0501010100-o        -1.148000e-05         2.078000e-05  
ID:0500010100-o        -3.110000e-06        -6.320000e-06  
ID:0602010100-o        -3.080000e-06        -1.133000e-05  
ID:0601010100-o         7.970000e-06        -2.137000e-05  
ID:0600010100-o        -2.200000e-07        -2.900000e-07  
ID:0702010100-o        -3.550000e-06        -1.123000e-05  
ID:0701010100-o         6.220000e-06        -2.377000e-05  
ID:0700010100-o        -2.300000e-07        -3.100000e-07

# Calculating magnetization

## Cartesian (with direction)

In [39]:
# calculating total magnetic moment by summing all
# (Total moment (of an orbital) * Direction of J (of the same orbital))
totmagnmomC = 0
for key in totmomentsD.keys():
    totmagnmomC += totmomentsD[key][0]*[x for x in dirofJD[key] if abs(x) > 0.9][0]
    print(totmomentsD[key][0], '*', 
          [x for x in dirofJD[key] if abs(x) > 0.9][0], '=', 
          totmomentsD[key][0]*[x for x in dirofJD[key] if abs(x) > 0.9][0])
    #    total magnetic moment J=L+S * orientation +/-1, wherever it is

print(f'\nTotal magnetic moment = {totmagnmomC}')

# getting unit cell volume in A^3 
ucv = findlinevaldict(lines, 'unit cell volume:')
ucvA = ucv[list(ucv.keys())[0]][0] / 1.8897259**3 # unit cell volume in A
print(f'Unit cell volume: {ucvA} A\N{SUPERSCRIPT THREE}')

# calculating magnetization in Tesla
magnetizationT = totmagnmomC/ucvA*11.654
print(f'Magnetization: {magnetizationT} T')

0.166403 * -1.0 = -0.166403
0.001853875 * -1.0 = -0.001853875
0.002103611 * -1.0 = -0.002103611
0.03702574 * -1.0 = -0.03702574
0.01078338 * -1.0 = -0.01078338
0.03702562 * -1.0 = -0.03702562
0.01078338 * -1.0 = -0.01078338
0.01127024 * -1.0 = -0.01127024
0.004421223 * -1.0 = -0.004421223
0.01127029 * -1.0 = -0.01127029
0.004421223 * -1.0 = -0.004421223
1.775499 * 1.0 = 1.775499
0.005087402 * 1.0 = 0.005087402
0.00581304 * 1.0 = 0.00581304
1.775499 * 1.0 = 1.775499
0.005087399 * 1.0 = 0.005087399
0.00581304 * 1.0 = 0.00581304

Total magnetic moment = 3.275437299
Unit cell volume: 56.01633474574038 A³
Magnetization: 0.6814431264703318 T


## Spin axis (no direction?)

In [40]:
# calculating total magnetic moment by summing all
# (Total moment (of an orbital) * Direction of J (of the same orbital))
totmagnmomSA = 0
for key in totmomentsD.keys():
    totmagnmomSA += totmomentsD[key][1]
    print(totmomentsD[key][1])
    #    total magnetic moment J=L+S * orientation +/-1, wherever it is

print(f'\nTotal magnetic moment = {totmagnmomSA}')

# getting unit cell volume in A^3 
ucv = findlinevaldict(lines, 'unit cell volume:')
ucvA = ucv[list(ucv.keys())[0]][0] / 1.8897259**3 # unit cell volume in A
print(f'Unit cell volume: {ucvA} A\N{SUPERSCRIPT THREE}')

# calculating magnetization in Tesla
magnetizationT = totmagnmomSA/ucvA*11.654
print(f'Magnetization: {magnetizationT} T')

0.166403
0.001853875
0.002103611
0.03702574
0.01078338
0.03702562
0.01078338
0.01127024
0.004421223
0.01127029
0.004421223
1.775499
0.005087402
0.00581304
1.775499
0.005087399
0.00581304

Total magnetic moment = 3.870160463
Unit cell volume: 56.01633474574038 A³
Magnetization: 0.805173173870176 T


# Results
<table border="1">
  <tr>
    <th rowspan = "2">Direction</th>
    <th colspan = "3">Cartesian</th>
    <th colspan = "3">Spin axis</th>
  </tr>
  <tr>
    <td><b>Total magnetic moment</b></td>
    <td><b>Unit cell volume (&#8491;<sup>3</sup>)</b></td>
    <td><b>Magnetization (T)</b></td>
    <td><b>Total magnetic moment</b></td>
    <td><b>Unit cell volume (&#8491;<sup>3</sup>)</b></td>
    <td><b>Magnetization (T)</b></td>
  </tr>
  <tr>
    <td><b>x</b></td>
    <td>3.272729207</td>
    <td>56.01633474574038</td>
    <td>0.6808797175234369</td>
    <td>3.870361919</td>
    <td>56.01633474574038</td>
    <td>0.8052150860772966</td>
  </tr>
  <tr>
    <td><b>z</b></td>
    <td>3.275437299</td>
    <td>56.01633474574038</td>
    <td>0.6814431264703318</td>
    <td>3.870160463</td>
    <td>56.01633474574038</td>
    <td>0.805173173870176</td>
  </tr>
</table>